packages/suite/src/components/wallet/WalletLayout/components/CoinjoinAccountDiscovery/useAccountLoadingProgress.ts

```diff
@@ -0,0 +1,87 @@
+/* eslint-disable no-nested-ternary */
+
+import { useEffect, useReducer } from 'react';
+import { ScanProgressInfo } from '@trezor/coinjoin/lib/types/backend';
+
+import { CoinjoinService } from 'src/services/coinjoin';
+import { selectSelectedAccount } from 'src/reducers/wallet/selectedAccountReducer';
+import { useSelector } from 'src/hooks/suite/useSelector';
+import { getAccountProgressHandle } from 'src/utils/wallet/coinjoinUtils';
+import { TranslationKey } from 'src/components/suite/Translation';
+
+const INIT_THRESHOLD = 0.05;
+const MEMPOOL_THRESHOLD = 0.85;
+
+type ProgressInfoReducerState = {
+    messageId?: TranslationKey;
+    progress: number;
+    stage?: 'block' | 'mempool';
+    outOf?: { current: number; total: number };
+};
+
+const progressInfoReducer = (
+    state: ProgressInfoReducerState,
+    info: ScanProgressInfo,
+): ProgressInfoReducerState => {
+    const outOf = !info.progress
+        ? info.stage === state.stage
+            ? state.outOf
+            : undefined
+        : info.stage === 'mempool'
+        ? info.progress
+        : {
+              current: info.progress.current - info.progress.from,
+              total: info.progress.to - info.progress.from,
+          };
+
+    const progress = outOf
+        ? info.stage === 'block'
+            ? INIT_THRESHOLD + (outOf.current / outOf.total) * (MEMPOOL_THRESHOLD - INIT_THRESHOLD)
+            : MEMPOOL_THRESHOLD + (outOf.current / outOf.total) * (1 - MEMPOOL_THRESHOLD)
+        : state.stage === 'block' && info.stage === 'mempool'
+        ? MEMPOOL_THRESHOLD
+        : info.stage === 'block'
+        ? INIT_THRESHOLD
+        : state.progress;
+
+    const messageId = `TR_COINJOIN_DISCOVERY_${info.stage === 'block' ? 'BLOCK' : 'MEMPOOL'}_${
+        outOf ? 'PROGRESS' : 'FETCHING'
+    }` as const;
+
+    return {
+        messageId,
+        progress,
+        stage: info.stage,
+        outOf,
+    };
+};
+
+export const useAccountLoadingProgress = () => {
+    const selectedAccount = useSelector(selectSelectedAccount);
+    const [progressInfo, dispatchProgressInfo] = useReducer(progressInfoReducer, {
+        progress: 0,
+    });
+
+    const { symbol: network, backendType } = selectedAccount || {};
+    const progressHandle = selectedAccount && getAccountProgressHandle(selectedAccount);
+
+    useEffect(() => {
+        if (!network || backendType !== 'coinjoin') {
+            return;
+        }
+
+        const api = CoinjoinService.getInstance(network);
+
+        if (!api) {
+            return;
+        }
+
+        api.backend.on(`progress-info/${progressHandle}`, dispatchProgressInfo);
```

Comment:
this can be `progress-info/undefined`?

Reply:
It can, but it shouldn't be ofc. I'll add the check.

Reply:
https://github.com/trezor/trezor-suite/pull/8826/commits/34146d67a5844e437f595b00e95286b449c51bba